File: packages/sdk-node/src/agents/healer.ts
/**
 * Self-healer agent - diagnoses and fixes failing tests
 */

import { Anthropic } from "@anthropic-ai/sdk";
import { ExecutionResult } from "../engine/executor.js";

export interface Diagnosis {
  issue_type:
    | "assertion_mismatch"
    | "timeout"
    | "network_error"
    | "syntax_error"
    | "logic_error"
    | "environment_issue"
    | "unknown";
  severity: "critical" | "high" | "medium" | "low";
  root_cause: string;
  affected_assertions: string[];
  recommendations: string[];
}

const client = new Anthropic();

/**
 * Self-healing agent that diagnoses and fixes test failures
 */
export class SelfHealer {
  /**
   * Diagnose a test failure
   */
  async diagnose(failure: ExecutionResult): Promise<Diagnosis> {
    const diagnosisPrompt = this.buildDiagnosisPrompt(failure);

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 2048,
      messages: [
        {
          role: "user",
          content: diagnosisPrompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      throw new Error("Unexpected response type from Claude");
    }

    const diagnosis = JSON.parse(content.text);

    return {
      issue_type: diagnosis.issue_type,
      severity: diagnosis.severity,
      root_cause: diagnosis.root_cause,
      affected_assertions: diagnosis.affected_assertions,
      recommendations: diagnosis.recommendations,
    };
  }

  /**
   * Heal a failing test by generating fixed code
   */
  async heal(diagnosis: Diagnosis, testCode: string): Promise<string> {
    const healingPrompt = this.buildHealingPrompt(diagnosis, testCode);

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 4096,
      messages: [
        {
          role: "user",
          content: healingPrompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      throw new Error("Unexpected response type from Claude");
    }

    // Extract code from markdown if present
    const text = content.text;
    const codeMatch = text.match(/```(?:typescript|javascript)?\n([\s\S]*?)\n```/);
    return codeMatch ? codeMatch[1] : text;
  }

  /**
   * Attempt automatic recovery
   */
  async recover(failure: ExecutionResult): Promise<{ recovered: boolean; fixedCode?: string }> {
    try {
      const diagnosis = await this.diagnose(failure);

      // Don't attempt recovery for critical issues
      if (diagnosis.severity === "critical") {
        return { recovered: false };
      }

      // Check if we have an original test code reference
      // In a real implementation, this would be tracked separately
      if (!failure.output) {
        return { recovered: false };
      }

      // Only attempt recovery for specific recoverable issues
      const recoverableIssues = [
        "assertion_mismatch",
        "timeout",
        "logic_error",
      ];

      if (!recoverableIssues.includes(diagnosis.issue_type)) {
        return { recovered: false };
      }

      // Attempt to generate fixed code
      const fixedCode = await this.heal(diagnosis, failure.output);

      return { recovered: true, fixedCode };
    } catch (error) {
      return { recovered: false };
    }
  }

  /**
   * Build diagnosis prompt
   */
  private buildDiagnosisPrompt(failure: ExecutionResult): string {
    return `Diagnose the following test failure:

EXECUTION STATUS: ${failure.status}
ASSERTIONS PASSED: ${failure.passed_count}
ASSERTIONS FAILED: ${failure.failed_count}
DURATION: ${failure.duration_ms}ms

TEST OUTPUT:
${failure.output}

${failure.error ? `ERROR MESSAGE:\n${failure.error}` : ""}

FAILED ASSERTIONS:
${failure.assertions
  .filter((a) => !a.passed)
  .map((a) => `- ${a.assertion}: ${a.error || "assertion failed"}`)
  .join("\n")}

Analyze the failure and return JSON diagnosis:
{
  "issue_type": "assertion_mismatch|timeout|network_error|syntax_error|logic_error|environment_issue|unknown",
  "severity": "critical|high|medium|low",
  "root_cause": "Detailed explanation of what went wrong",
  "affected_assertions": ["list of affected assertion IDs"],
  "recommendations": [
    "First recommendation to fix the issue",
    "Second recommendation",
    "..."
  ]
}`;
  }

  /**
   * Build healing prompt
   */
  private buildHealingPrompt(diagnosis: Diagnosis, testCode: string): string {
    return `Fix the following failing test code based on the diagnosis:

DIAGNOSIS:
Issue Type: ${diagnosis.issue_type}
Severity: ${diagnosis.severity}
Root Cause: ${diagnosis.root_cause}

RECOMMENDATIONS:
${diagnosis.recommendations.map((r, i) => `${i + 1}. ${r}`).join("\n")}

ORIGINAL TEST CODE:
\`\`\`typescript
${testCode}
\`\`\`

FAILING ASSERTIONS:
${diagnosis.affected_assertions.join("\n")}

Generate fixed test code that:
1. Addresses the root cause
2. Implements the recommendations
3. Maintains all original test intent
4. Includes improved error handling
5. Adds appropriate logging/debugging

Return ONLY the complete, fixed Playwright test code in a TypeScript code block.`;
  }

  /**
   * Suggest alternative test approaches
   */
  async suggestAlternatives(
    failure: ExecutionResult
  ): Promise<string[]> {
    const prompt = `Given this test failure, suggest alternative testing approaches:

STATUS: ${failure.status}
ERROR: ${failure.error}
OUTPUT: ${failure.output}

Return a JSON array of alternative approaches:
["Approach 1", "Approach 2", "Approach 3"]`;

    try {
      const response = await client.messages.create({
        model: "claude-opus-4-6",
        max_tokens: 1024,
        messages: [
          {
            role: "user",
            content: prompt,
          },
        ],
      });

      const content = response.content[0];
      if (content.type !== "text") {
        return [];
      }

      const alternatives = JSON.parse(content.text);
      return Array.isArray(alternatives) ? alternatives : [];
    } catch {
      return [];
    }
  }
}

/**
 * Healing strategy pattern
 */
export interface HealingStrategy {
  canHandle(diagnosis: Diagnosis): boolean;
  apply(testCode: string, diagnosis: Diagnosis): Promise<string>;
}

/**
 * Assertion mismatch healing strategy
 */
export class AssertionMismatchHealer implements HealingStrategy {
  canHandle(diagnosis: Diagnosis): boolean {
    return diagnosis.issue_type === "assertion_mismatch";
  }

  async apply(testCode: string, diagnosis: Diagnosis): Promise<string> {
    const prompt = `Fix assertion mismatches in this test code:

ROOT CAUSE: ${diagnosis.root_cause}

TEST CODE:
\`\`\`typescript
${testCode}
\`\`\`

Update the assertions to match the actual behavior. Return fixed code.`;

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 2048,
      messages: [
        {
          role: "user",
          content: prompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      return testCode;
    }

    const match = content.text.match(/```(?:typescript)?\n([\s\S]*?)\n```/);
    return match ? match[1] : testCode;
  }
}

/**
 * Timeout healing strategy
 */
export class TimeoutHealer implements HealingStrategy {
  canHandle(diagnosis: Diagnosis): boolean {
    return diagnosis.issue_type === "timeout";
  }

  async apply(testCode: string, diagnosis: Diagnosis): Promise<string> {
    // Add longer timeouts and wait strategies
    let fixed = testCode;

    // Increase test timeout
    fixed = fixed.replace(/test\.setTimeout\(\d+\)/, "test.setTimeout(60000)");
    if (!fixed.includes("test.setTimeout")) {
      fixed = `test.setTimeout(60000);\n\n${fixed}`;
    }

    // Add explicit waits
    fixed = fixed.replace(
      /expect\(/g,
      "await page.waitForLoadState('networkidle');\nexpect("
    );

    return fixed;
  }
}

/**
 * Environment issue healing strategy
 */
export class EnvironmentIssueHealer implements HealingStrategy {
  canHandle(diagnosis: Diagnosis): boolean {
    return diagnosis.issue_type === "environment_issue";
  }

  async apply(testCode: string, diagnosis: Diagnosis): Promise<string> {
    const prompt = `Fix environment-related issues in this test:

ISSUES: ${diagnosis.root_cause}

TEST CODE:
\`\`\`typescript
${testCode}
\`\`\`

Add environment variable validation, retry logic, and fallback values. Return fixed code.`;

    const response = await client.messages.create({
      model: "claude-opus-4-6",
      max_tokens: 2048,
      messages: [
        {
          role: "user",
          content: prompt,
        },
      ],
    });

    const content = response.content[0];
    if (content.type !== "text") {
      return testCode;
    }

    const match = content.text.match(/```(?:typescript)?\n([\s\S]*?)\n```/);
    return match ? match[1] : testCode;
  }
}

File: packages/sdk-python/rigour_qa/agents/healer.py
"""Self-healing agent for diagnosing and fixing test failures."""

from typing import Optional

from anthropic import Anthropic

from rigour_qa.engine.types import Diagnosis, ExecutionResult


class SelfHealer:
    """Diagnoses test failures and generates healing suggestions."""

    def __init__(self, api_key: Optional[str] = None):
        """Initialize healer with Anthropic client."""
        self.client = Anthropic(api_key=api_key)
        self.failure_history: dict[str, list[ExecutionResult]] = {}

    def diagnose(self, failure: ExecutionResult) -> Diagnosis:
        """Analyze a test failure and provide diagnosis."""

        prompt = f"""Analyze this test failure and provide a root cause diagnosis:

Test ID: {failure.plan_id}
Status: {failure.status}
Error: {failure.error_message}
Stderr: {failure.stderr[:500]}

Determine:
1. Root cause (what went wrong)
2. Failure type (assertion, timeout, network, syntax, other)
3. Is it likely flaky? (intermittent/timing-dependent)
4. Suggested fixes

Return JSON: {{
    "root_cause": "...",
    "failure_type": "...",
    "is_flaky": bool,
    "suggested_fixes": ["fix1", "fix2", "fix3"]
}}"""

        message = self.client.messages.create(
            model="claude-opus-4-6",
            max_tokens=600,
            messages=[{"role": "user", "content": prompt}]
        )

        response_text = message.content[0].text

        try:
            import json
            import re
            json_match = re.search(r'\{.*\}', response_text, re.DOTALL)
            if json_match:
                diagnosis_dict = json.loads(json_match.group())
            else:
                diagnosis_dict = json.loads(response_text)
        except json.JSONDecodeError:
            diagnosis_dict = {
                "root_cause": "Unknown",
                "failure_type": "error",
                "is_flaky": False,
                "suggested_fixes": ["Review test code and error message"],
            }

        return Diagnosis(
            test_id=failure.plan_id,
            root_cause=diagnosis_dict.get("root_cause", "Unknown"),
            failure_type=diagnosis_dict.get("failure_type", "error"),
            is_flaky=diagnosis_dict.get("is_flaky", False),
            suggested_fixes=diagnosis_dict.get("suggested_fixes", []),
        )

    def heal(self, diagnosis: Diagnosis, test_code: str) -> str:
        """Generate healed test code based on diagnosis."""

        prompt = f"""Fix this test code based on the diagnosis:

Root Cause: {diagnosis.root_cause}
Failure Type: {diagnosis.failure_type}
Is Flaky: {diagnosis.is_flaky}
Suggested Fixes: {', '.join(diagnosis.suggested_fixes)}

Original Test Code:
```python
{test_code}
```

Generate corrected test code that:
1. Addresses the identified root cause
2. Adds error handling if needed
3. Adds retries if test is flaky
4. Improves assertions
5. Is production-quality

Return ONLY the corrected Python code (no markdown, no explanation)."""

        message = self.client.messages.create(
            model="claude-opus-4-6",
            max_tokens=3000,
            messages=[{"role": "user", "content": prompt}]
        )

        response_text = message.content[0].text
        healed_code = response_text.strip()

        if healed_code.startswith("```"):
            import re
            code_match = re.search(r'```python\n(.*?)\n```', healed_code, re.DOTALL)
            if code_match:
                healed_code = code_match.group(1)

        return healed_code

    def is_flaky(self, test_id: str, history: list[ExecutionResult]) -> bool:
        """Determine if a test is flaky based on execution history."""

        if test_id not in self.failure_history:
            self.failure_history[test_id] = []

        self.failure_history[test_id].extend(history)
        recent = self.failure_history[test_id][-10:]

        if len(recent) < 3:
            return False

        passed_count = sum(1 for r in recent if r.passed)
        failed_count = len(recent) - passed_count

        if failed_count == 0 or passed_count == 0:
            return False

        pass_rate = passed_count / len(recent)
        flaky = 0.3 <= pass_rate <= 0.7

        if flaky and failed_count >= 2:
            return True

        return False

    def generate_healing_plan(
        self,
        failures: list[ExecutionResult],
        test_codes: dict[str, str],
    ) -> dict[str, str]:
        """Generate healing code for multiple failures."""

        healed = {}

        for failure in failures:
            if failure.plan_id not in test_codes:
                continue

            diagnosis = self.diagnose(failure)
            original_code = test_codes[failure.plan_id]
            healed_code = self.heal(diagnosis, original_code)
            healed[failure.plan_id] = healed_code

        return healed
